File: src/main/darktrade.js
'use strict';

/**
 * 东方财富暗盘资金。
 *
 * 口径与字段映射对齐 gupiao_ztfx 的 `src/dark_trade/eastmoney.py`：
 * 接口返回的是「某个交易日的暗盘排行榜」，按暗盘资金降序分页，
 * 所以这里翻页收集，直到自选股都命中或翻到页数上限为止。
 *
 * 返回体形如：
 *   { errid: 0, "1": "20260817", "2": 总条数, data: [ { "4": 代码, "6": 暗盘资金, ... } ] }
 */

const ENDPOINT = 'https://quotederivates.eastmoney.com/datacenter/darktrade';
const PAGE_SIZE = 100;
const MAX_PAGES = 50; // 5000 条，足够覆盖一天有暗盘成交的个股
const PRICE_DIVISOR = 1000; // 接口里的价格放大了 1000 倍
const CACHE_TTL_MS = 10 * 60 * 1000; // 日频数据，不必跟行情同频刷新
const REQUEST_TIMEOUT_MS = 10000;

// 接口字段是数字键，含义按 gupiao_ztfx 的映射表。
const F = {
  MARKET: '3',
  CODE: '4',
  DARK_FUND: '6',
  OPEN_FUND: '7',
  MAIN_NET_INFLOW: '8',
  ACTIVITY: '11',
  PRICE: '13',
  CHANGE_PERCENT: '14',
  NAME: '16',
  RANK: '21',
};

const num = (v) => {
  const n = Number(v);
  return Number.isFinite(n) ? n : null;
};

/** 接口要 YYYYMMDD；本地日期即可，返回体里会给出实际交易日。 */
function todayStamp(now = new Date()) {
  const pad = (n) => String(n).padStart(2, '0');
  return `${now.getFullYear()}${pad(now.getMonth() + 1)}${pad(now.getDate())}`;
}

function normalizeRow(row) {
  if (!row || typeof row !== 'object') return null;
  const digits = String(row[F.CODE] ?? '').replace(/\D/g, '');
  if (!digits) return null;
  const price = num(row[F.PRICE]);
  return {
    code: digits.padStart(6, '0'),
    name: String(row[F.NAME] ?? '').trim(),
    market: num(row[F.MARKET]) === 1 ? 'sh' : 'sz',
    darkFund: num(row[F.DARK_FUND]) ?? 0, // 单位：元
    openFund: num(row[F.OPEN_FUND]) ?? 0,
    mainNetInflow: num(row[F.MAIN_NET_INFLOW]) ?? 0,
    activity: num(row[F.ACTIVITY]) ?? 0,
    price: price === null ? null : price / PRICE_DIVISOR,
    changePercent: num(row[F.CHANGE_PERCENT]),
    rank: num(row[F.RANK]),
  };
}

async function requestPage(dateStamp, page) {
  const params = new URLSearchParams({
    version: '101',
    cver: '100',
    date: dateStamp,
    StartPage: String(page),
    NumPerPage: String(PAGE_SIZE),
    sortflag: '6',
    desc: '1',
    market: '',
    datetype: '',
  });
  const res = await fetch(`${ENDPOINT}?${params}`, {
    headers: {
      'User-Agent': 'Mozilla/5.0',
      Accept: 'application/json,text/plain,*/*',
      Referer: 'https://emrnweb.eastmoney.com/graymarket/rankList',
      rnProjectId: 'emrn.GrayMarketRank',
    },
    signal: AbortSignal.timeout(REQUEST_TIMEOUT_MS),
  });
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  const payload = await res.json();
  if (!payload || typeof payload !== 'object') throw new Error('返回格式异常');
  if (num(payload.errid) !== 0) {
    throw new Error(String(payload.errmsg || '接口返回错误').slice(0, 60));
  }
  return payload;
}

/** 把 "20260817" 变成 "2026-08-17"，拿不到就返回空串。 */
function formatDate(value) {
  const text = String(value ?? '').replace(/-/g, '');
  return /^\d{8}$/.test(text) ? `${text.slice(0, 4)}-${text.slice(4, 6)}-${text.slice(6)}` : '';
}

let cache = null; // { at, date, byCode, error, codes }

/**
 * 取指定代码的暗盘资金。
 * @param {string[]} codes 6 位股票代码
 * @returns {Promise<{ date: string, byCode: Map<string, object>, error: string|null }>}
 */
async function getDarkTrade(codes, { now = Date.now() } = {}) {
  const wanted = new Set(codes);
  const cacheUsable =
    cache &&
    now - cache.at < CACHE_TTL_MS &&
    // 自选变了且新代码上次没覆盖到时，重新拉一次
    [...wanted].every((code) => cache.codes.has(code));
  if (cacheUsable) return { date: cache.date, byCode: cache.byCode, error: cache.error };

  const byCode = new Map();
  let date = '';
  let error = null;

  try {
    const dateStamp = todayStamp();
    const first = await requestPage(dateStamp, 1);
    date = formatDate(first['1']);
    const total = num(first['2']) ?? 0;
    const pages = Math.min(MAX_PAGES, Math.max(1, Math.ceil(total / PAGE_SIZE)));

    const absorb = (payload) => {
      for (const raw of payload?.data || []) {
        const row = normalizeRow(raw);
        if (row) byCode.set(row.code, row);
      }
    };
    absorb(first);

    for (let page = 2; page <= pages; page++) {
      // 自选股全部命中就不必继续翻页了。
      if ([...wanted].every((code) => byCode.has(code))) break;
      absorb(await requestPage(dateStamp, page));
    }
  } catch (err) {
    error = err.name === 'TimeoutError' ? '请求超时' : err.message;
  }

  cache = { at: now, date, byCode, error, codes: wanted };
  return { date, byCode, error };
}

/** 仅供测试：清掉进程内缓存。 */
function resetCache() {
  cache = null;
}

module.exports = { getDarkTrade, normalizeRow, formatDate, todayStamp, resetCache, PRICE_DIVISOR };

File: stockwidget/darktrade.py
"""东方财富暗盘资金。

口径与字段映射对齐 ``gupiao_ztfx`` 的 ``src/dark_trade/eastmoney.py``：
接口返回的是「某个交易日的暗盘排行榜」，按暗盘资金降序分页，
所以这里翻页收集，直到自选股都命中或翻到页数上限为止。

返回体形如::

    {"errid": 0, "1": "20260817", "2": 总条数, "data": [{"4": 代码, "6": 暗盘资金, ...}]}
"""

from __future__ import annotations

import math
import time
from dataclasses import dataclass
from datetime import date
from typing import Any, Iterable

import requests

from .providers.base import USER_AGENT, describe_error

ENDPOINT = "https://quotederivates.eastmoney.com/datacenter/darktrade"
PAGE_SIZE = 100
# 东方财富的榜单已经可能超过 5000 条，50 页会漏掉尾部个股。
# 仍保留一个宽松的安全上限，避免接口总数异常时无限翻页。
MAX_PAGES = 100
PRICE_DIVISOR = 1000  # 接口里的价格放大了 1000 倍
CACHE_TTL_SECONDS = 10 * 60  # 日频数据，不必跟行情同频刷新
REQUEST_TIMEOUT = 10

# 接口字段是数字键，含义按 gupiao_ztfx 的映射表。
F_MARKET = "3"
F_CODE = "4"
F_DARK_FUND = "6"
F_OPEN_FUND = "7"
F_MAIN_NET_INFLOW = "8"
F_ACTIVITY = "11"
F_PRICE = "13"
F_CHANGE_PERCENT = "14"
F_NAME = "16"
F_RANK = "21"


def _number(value: Any, default: float | None = None) -> float | None:
    try:
        result = float(value)
    except (TypeError, ValueError):
        return default
    return result if math.isfinite(result) else default


@dataclass
class DarkRow:
    code: str
    name: str
    market: str
    dark_fund: float  # 单位：元
    open_fund: float
    main_net_inflow: float
    activity: float
    price: float | None
    change_percent: float | None
    rank: int | None


@dataclass
class DarkResult:
    trade_date: str = ""
    by_code: dict[str, DarkRow] = None  # type: ignore[assignment]
    error: str | None = None

    def __post_init__(self) -> None:
        if self.by_code is None:
            self.by_code = {}


def normalize_row(row: Any) -> DarkRow | None:
    if not isinstance(row, dict):
        return None
    digits = "".join(ch for ch in str(row.get(F_CODE) or "") if ch.isdigit())
    if not digits:
        return None
    price = _number(row.get(F_PRICE))
    return DarkRow(
        code=digits.zfill(6),
        name=str(row.get(F_NAME) or "").strip(),
        market="sh" if _number(row.get(F_MARKET)) == 1 else "sz",
        dark_fund=_number(row.get(F_DARK_FUND), 0.0) or 0.0,
        open_fund=_number(row.get(F_OPEN_FUND), 0.0) or 0.0,
        main_net_inflow=_number(row.get(F_MAIN_NET_INFLOW), 0.0) or 0.0,
        activity=_number(row.get(F_ACTIVITY), 0.0) or 0.0,
        price=None if price is None else price / PRICE_DIVISOR,
        change_percent=_number(row.get(F_CHANGE_PERCENT)),
        rank=int(_number(row.get(F_RANK), 0) or 0) or None,
    )


def format_date(value: Any) -> str:
    """``20260817`` -> ``2026-08-17``，拿不到就返回空串。"""
    text = str(value or "").replace("-", "")
    return f"{text[:4]}-{text[4:6]}-{text[6:]}" if len(text) == 8 and text.isdigit() else ""


class DarkTradeClient:
    """带缓存的暗盘资金查询。"""

    def __init__(self, session: requests.Session | None = None) -> None:
        self.session = session or requests.Session()
        self._cache: DarkResult | None = None
        self._cached_at = 0.0
        self._cached_codes: set[str] = set()

    def _request_page(self, date_stamp: str, page: int) -> dict:
        response = self.session.get(
            ENDPOINT,
            params={
                "version": 101,
                "cver": 100,
                "date": date_stamp,
                "StartPage": page,
                "NumPerPage": PAGE_SIZE,
                "sortflag": 6,
                "desc": 1,
                "market": "",
                "datetype": "",
            },
            headers={
                "User-Agent": USER_AGENT,
                "Accept": "application/json,text/plain,*/*",
                "Referer": "https://emrnweb.eastmoney.com/graymarket/rankList",
                "rnProjectId": "emrn.GrayMarketRank",
            },
            timeout=REQUEST_TIMEOUT,
        )
        response.raise_for_status()
        payload = response.json()
        if not isinstance(payload, dict):
            raise ValueError("返回格式异常")
        # 注意不能写成 `_number(...) or -1`：errid 正常就是 0，会被 or 吃掉。
        errid = _number(payload.get("errid"))
        if errid is None or int(errid) != 0:
            raise ValueError(str(payload.get("errmsg") or "接口返回错误")[:60])
        return payload

    def fetch(self, codes: Iterable[str], now: float | None = None) -> DarkResult:
        """取指定 6 位代码的暗盘资金；失败时返回带 error 的空结果。"""
        wanted = set(codes)
        now = time.time() if now is None else now
        if (
            self._cache is not None
            and now - self._cached_at < CACHE_TTL_SECONDS
            # 自选变了且新代码上次没覆盖到时，重新拉一次
            and wanted <= self._cached_codes
        ):
            return self._cache

        result = DarkResult()
        try:
            date_stamp = date.today().strftime("%Y%m%d")
            first = self._request_page(date_stamp, 1)
            result.trade_date = format_date(first.get("1"))
            total = int(_number(first.get("2"), 0) or 0)
            pages = min(MAX_PAGES, max(1, math.ceil(total / PAGE_SIZE)))

            def absorb(payload: dict) -> None:
                for raw in payload.get("data") or []:
                    if row := normalize_row(raw):
                        result.by_code[row.code] = row

            absorb(first)
            for page in range(2, pages + 1):
                if wanted <= result.by_code.keys():  # 自选全部命中就不必继续翻页
                    break
                absorb(self._request_page(date_stamp, page))
        except Exception as exc:
            result.error = describe_error(exc)

        self._cache, self._cached_at, self._cached_codes = result, now, wanted
        return result

    def reset_cache(self) -> None:
        """仅供测试。"""
        self._cache = None
        self._cached_codes = set()
